### PR TITLE
Add utils package to generate id tokens 

### PR DESCRIPTION
**Pull request by deadlycoconuts**
## Context
As of present, authentication is performed by the Merlin and Turing SDKs using Google OAuth 2.0 access tokens (generated after authentication with Google's OAuth server) that are added to the headers of requests sent to the Turing API server. These tokens are then used by an internal proxy which will validate the access token and then retrieve the user profile information from the Google API. 

However, in keeping with the move of the Google Identity Services client library from OAuth 2.0 access tokens to Google OpenID Connect (OIDC) identity tokens, we are similarly replacing the use of the former for the latter. These tokens generated would similarly be added to the headers of requests sent to the Turing API server to be validated internally. 

This PR serves to introduce the new authentication methods by making the Google API Python client generate id token credentials instead of OAuth2 ones. The new authentication methods will then be imported by both the Merlin and Turing SDKs to perform authentication of the user.

## Changes
The bulk of the code introduced are almost entirely drawn from Apache's Airflow Google [utils library](https://github.com/apache/airflow/blob/main/airflow/providers/google/common/utils/id_token_credentials.py) that authenticates an SDK user and generates Google OIDC identity token credentials, from three potential sources (in decreasing priority),
1. Credential files paths specified with the `GOOGLE_APPLICATION_CREDENTIALS` environment variable
2. GCloud SDK credentials (includes application default credentials)
3. Google Compute Engine credentials

Regardless of where the credentials are found, note in particular the difference between generating identity tokens from a user account and from a service account as Google does not allow the generation of identity tokens with user accounts but only from service accounts.

When creating id tokens from **service accounts**, we simply use the methods provided by the Google API python client library to generate them in a `google.oauth2.service_account.IDTokenCredentials` object. 

However, when creating id tokens from user accounts, since this is not immediately possible, we're initialising a `IDTokenCredentialsAdapter` wrapper around the OAuth2 credentials, which 'mocks' the `IDTokenCrendtials` class by having a `None` id token. When the credentials are refreshed, the `id_token` field will automatically be populated with the default `client_id` for whichever credentials are used:

Google Cloud default applications credentials `client_id` and `aud` value:
`764086051850-6qr4p6gpi6hn506pt8ejuq83di341hur.apps.googleusercontent.com`

Google Cloud SDK `client_id` and `aud` value:
`32555940559.apps.googleusercontent.com`

```python
if credential_type == _AUTHORIZED_USER_TYPE:
    current_credentials = oauth2_credentials.Credentials.from_authorized_user_info(
        info,
    )
    current_credentials = IDTokenCredentialsAdapter(credentials=current_credentials)

    return current_credentials

elif credential_type == _SERVICE_ACCOUNT_TYPE:
    try:
        return service_account.IDTokenCredentials.from_service_account_info(
            info, target_audience=target_audience
        )
    except ValueError:
        raise exceptions.DefaultCredentialsError(
            f"Failed to load service account credentials from {filename}"
        )
```

## Intended Usage
Since the authentication methods will be bundled in a `caraml-auth-google` package, that contains the `caraml_auth` module, the Merlin and Turing SDKs would have to import and use the authentication methods introduced in a way that resembles:

```python
from caraml_auth.id_token_credentials import get_default_id_token_credentials

credentials = get_default_id_token_credentials(target_audience="turing-sdk.caraml")
credentials.refresh(Request())
```

cc. @mbruner

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,4 @@
-# caraml-sdk
\ No newline at end of file
+# CaraML SDK
+
+Python SDK for interacting with the Merlin and Turing SDKs of CaraML.
```
Comment (krithika369) [inline README.md]: Nit: The `caraml-sdk` is not yet built, we're only building the collecting of common dependencies under it now. So, shall we rephrase as:

> Common Python SDK dependencies for CaraML components

**Comment by krithika369** [comment]
> with the move to decouple authentication from authorization

What does this mean?

File: packages/caraml-auth-google/.gitignore [hunk 2]
```diff
@@ -0,0 +1,5 @@
+build/
```
Comment (krithika369) [inline packages/caraml-auth-google/.gitignore]: Can these be moved to a root level `.gitignore`, replacing `caraml_auth_google.egg-info.egg-info/` with `*.egg-info` ?

File: packages/caraml-auth-google/README.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# CaraML Google Authentication
+
+Utility package containing functions to authenticate users of the Merlin and Turing SDKs of CaraML.
```
Comment (krithika369) [inline packages/caraml-auth-google/README.md]: Nit: Might be better to say "authenticate users of the Python SDKs of CaraML." than explicitly say Turing and Merlin. There is a chance, at some point, we could integrate other components (like DAP?) and right now, from this repo's POV, there is no explicit connection to Turing/Merlin with this repo.

File: packages/caraml-auth-google/requirements.dev.txt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+black==22.6.0
+setuptools>=21.0.0
+wheel
+twine
+pytest
```
Comment (krithika369) [inline packages/caraml-auth-google/requirements.dev.txt:5]: We don't have any test at the moment :)

But can we add it in? https://github.com/apache/airflow/blob/9ab1a6a3e70b32a3cddddf0adede5d2f3f7e29ea/tests/providers/google/common/utils/test_id_token_credentials.py

File: packages/caraml-auth-google/README.md (hunk 3 above)
Comment (krithika369) [reply packages/caraml-auth-google/README.md]: Also, some simple dev notes on installation (as of now, using the git repo ref), tests, etc. will be useful.

**Comment by deadlycoconuts** [comment]
> > with the move to decouple authentication from authorization
> 
> What does this mean?

Ah good catch; on my mind I was referring to the Google Identity Services library's move from access tokens to OIDC as one that decouples authentication from authorisation but on reading that sentence, I don't think it's even relevant given that the rest of the MR description is about our changes to OIDC instead of theirs.

File: packages/caraml-auth-google/version.py [hunk 5]
```diff
@@ -0,0 +1,2 @@
+# Do not update this field manually; this value gets updated automatically during build time
+VERSION = "0.0.0"
```
Comment (krithika369) [inline packages/caraml-auth-google/version.py]: We don't have any CI on this yet - shall we set it to `0.0.1` and come back to update this when we add the publish?

File: packages/caraml-auth-google/.gitignore (hunk 2 above)
Comment (deadlycoconuts) [reply packages/caraml-auth-google/.gitignore]: Hmm oh yes good point, will do that.

File: packages/caraml-auth-google/version.py (hunk 5 above)
Comment (deadlycoconuts) [reply packages/caraml-auth-google/version.py]: Alright, sounds good!

File: packages/caraml-auth-google/requirements.dev.txt (hunk 4 above)
Comment (deadlycoconuts) [reply packages/caraml-auth-google/requirements.dev.txt:5]: Okay I've added the test suite as well as a makefile, and a CI job to run it :)

File: packages/caraml-auth-google/README.md (hunk 3 above)
Comment (deadlycoconuts) [reply packages/caraml-auth-google/README.md]: Ok I've included additional dev docs on the usage and testing!

**Comment by deadlycoconuts** [comment]
Thanks for the quick comments! I've added quite a number of other boilerplate stuff to the repo like a CI workflow and a Makefile and other tests. You might want to take a brief look at those stuff too.

File: .github/workflows/caraml-sdk.yaml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        uses: conda-incubator/setup-miniconda@v2
+        with:
+          auto-update-conda: true
+
+      - name: Cache Conda environment
+        uses: actions/cache@v2
+        with:
+          path: engines/pyfunc-ensembler-service/env
+          key: |
+            conda-${{ hashFiles('engines/pyfunc-ensembler-service/env-${{ matrix.python-version }}.yaml') }}-${{ hashFiles('engines/pyfunc-ensembler-service/requirements.txt') }}-${{ hashFiles('engines/pyfunc-ensembler-service/requirements.dev.txt') }}
```
Comment (krithika369) [inline .github/workflows/caraml-sdk.yaml]: Copy-paste typo with: `pyfunc-ensembler-service`

File: .github/workflows/caraml-sdk.yaml [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Setup Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Setup Conda
+        uses: conda-incubator/setup-miniconda@v2
```
Comment (krithika369) [inline .github/workflows/caraml-sdk.yaml:45]: Why do we need Conda?

File: .github/workflows/caraml-sdk.yaml [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+      - name: Cache Conda environment
+        uses: actions/cache@v2
+        with:
+          path: engines/pyfunc-ensembler-service/env
+          key: |
+            conda-${{ hashFiles('engines/pyfunc-ensembler-service/env-${{ matrix.python-version }}.yaml') }}-${{ hashFiles('engines/pyfunc-ensembler-service/requirements.txt') }}-${{ hashFiles('engines/pyfunc-ensembler-service/requirements.dev.txt') }}
+          restore-keys: conda-
+
+      - name: Run Tests
+        working-directory: engines/pyfunc-ensembler-service
```
Comment (krithika369) [inline .github/workflows/caraml-sdk.yaml]: Also here.. Need to replace dir

File: packages/caraml-auth-google/README.md [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+functions contained in this package requires installing/specifying this package as a Git (VCS) dependency:
+
+To install via `pip`:
+```shell
+pip install git+https://github.com/caraml-dev/caraml-sdk.git@main#egg=caraml-auth-google&subdirectory=packages/caraml-auth-google
+```
+
+To specify this in a `requirements.txt` file:
+```txt
+# Other dependencies...
```
Comment (krithika369) [inline packages/caraml-auth-google/README.md]: What's this # Other dependencies ?

**Review by krithika369 — APPROVED** [review]
Left some comments, the rest LGTM!

File: .github/workflows/caraml-sdk.yaml (hunk 7 above)
Comment (deadlycoconuts) [reply .github/workflows/caraml-sdk.yaml:45]: Well it's just to install all the dependencies needed. But I guess with disposable runners we don't really need them 🤔

Comment (krithika369) [reply .github/workflows/caraml-sdk.yaml:45]: Ah ok.. Fair enough. I missed that the pip install in covered by this.

File: packages/caraml-auth-google/README.md (hunk 9 above)
Comment (deadlycoconuts) [reply packages/caraml-auth-google/README.md]: They are just meant to refer to other SDK dependencies. But I guess replacing that line with just
```txt
caraml-auth-google @ git+https://github.com/caraml-dev/caraml-sdk.git@main#egg=caraml-auth-google&subdirectory=packages/caraml-auth-google
```
is less confusing

**Comment by deadlycoconuts** [comment]
Thanks a lot for the comments @krithika369, I'll be merging this now! 🚀

**Pull request merged**